### ISSUE TITLE
List of most recently used systems for viewer filling in improperly

### ISSUE THREAD

**Issue by MerlinYoda** [ultravnc/UltraVNC#89]
The drop-down menu listing the most recent systems that were connected to does not populate properly and will lead to duplicates after enough entries are added. 

The issue is detailed in the forums when I first discovered it here: https://forum.uvnc.com/viewtopic.php?t=37670 
A shorter summary of it once the issue persisted between versions can be found here as well: https://forum.uvnc.com/viewtopic.php?t=37727

A patch for the file containing the code for responsible for generating and maintaining the listing that should resolve this issue (as well as slightly increase the efficiency of maintaining the listing) is attached as follows: [MRU.cpp.patch](https://github.com/ultravnc/UltraVNC/files/11627906/MRU.cpp.patch)

**Comment by RudiDeVos** [comment]
Thanks, patch wil be added

**Comment by RudiDeVos** [comment]
MRU patch added.

Changes, functions require TCHAR *, not TCHAR or CONST TCHAR*

_tcschr((TCHAR*)RESERVED_CHARS, firstUnusedId)
_tcscmp(&firstUnusedId, &LAST_USEABLE_ID)
_tcscmp(&firstUnusedId, (TCHAR*)LAST_USEABLE_ID)

https://github.com/ultravnc/UltraVNC/commit/9ca3d1b9f0493787eefc2768f7b0f0babddf6e06

Please verify if it's working as expected

**Comment by MerlinYoda** [comment]
So, I downloaded Visual Studio 2019 and, after wresting with some settings, I was able to compile the current main branch version of the viewer (fun fact: apparently having software execution restrictions set on the current user's temp folder doesn't go over well with the build process). While testing it, I got some, well, odd behavior. Mostly the viewer would crash (probably from a segmentation fault) as soon as you tried to connect to a host and it would also drop entries from the listing. However, with a small tweak, I was able to get to work just fine.

On line 88, change: 
_tcscmp(&firstUnusedId, (TCHAR*)LAST_USEABLE_ID)

to: 
_tcscmp(&firstUnusedId, &LAST_USEABLE_ID)

because LAST_USABLE_ID is a TCHAR and we need its address to pass the proper pointer value.

Also, on line 85, the (TCHAR *) cast on RESERVED_CHARS is redundant since it's already that type but it's not really hurting anything being there either.

**Comment by RudiDeVos** [comment]
Correct, commited before testing...

**Comment by Neustradamus** [comment]
@MerlinYoda: What do you think about current code?
It is good?

**Comment by MerlinYoda** [comment]
It's a lot better but there's still something not quite right. If you directly select one of the existing entries from the list it then will work as expected. Also, if you enter a value that is not listed, then it will add an entry as expected. 

However, there's an odd bug where, if you type in the one of the server entries that is already in the list instead of selecting it from the listing and connect to it, it will actually remove the letter it is indexed to from the index listing and simply duplicate whatever the first entry was, causing the first entry it to display twice in the MRU listing on the next run. However, it will not actually drop the key/value pair for that listing from the file. If you type in the same value as before and connect again, it will add its index back to the front of the list again with the duplicated indexes remaining in the list and the key-value pair listing will not the pushed to the bottom of the section like it usually does unless you select it from the listing on a later run.

Even stranger is that, in some circumstances, if you select or type in one of those duplicated entries after the previously missing index is re-inserted, it uses a _completely new index _ as if it wasn't already in the listing and it drops the old key/value pair while still keeping the old index in the index listing (which, does at least address the earlier duplication issue). The old index will be reused eventually after entering enough entries that are not in the listing. Although, the duplicated indexes will always persist and eventually cause newer entries in the listing to be overwritten because a duplicated index just happened to fall at the end of the list.

**Comment by MerlinYoda** [comment]
I figured out what the issue was. Apparently the line of code responsible for setting the first item in the index to the key of the matching value got dropped from the file. I added it back in and it works perfectly fine now. Here's the patch with that change (plus a bit of formatting cleanup): 
[MRU.cpp-202306231212.patch](https://github.com/ultravnc/UltraVNC/files/11850233/MRU.cpp-202306231212.patch)
